packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+    _scrollControllerX.dispose();
+    _scrollControllerY.dispose();
+    _selectedEventSubscription.cancel();
+    // TODO(kenz): dispose [_controller] here.
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Container(
-      color: Colors.black26,
-      child: const Center(
-        child: Text('TODO Flame Chart'),
-      ),
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return Scrollbar(
+          child: SingleChildScrollView(
+            controller: _scrollControllerX,
+            scrollDirection: Axis.horizontal,
+            child: Scrollbar(
+              child: SingleChildScrollView(
+                controller: _scrollControllerY,
+                scrollDirection: Axis.vertical,
+                child: _flameChartBody(constraints),
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  Widget _flameChartBody(BoxConstraints constraints) {
+    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
+    final height = math.max(constraints.maxHeight, widget.height);
+
+    return Stack(
+      children: [
+        Container(
+          width: width,
+          height: height,
+        ),
+        ..._nodesInViewport(constraints), // pick what to show
+      ],
+    );
+  }
+
+  List<FlameChartNode> _nodesInViewport(BoxConstraints constraints) {
+    // TODO(kenz): is creating all the FlameChartNode objects expensive even if
+    // we won't add them to the view? We create all the FlameChartNode objects
+    // and place them in FlameChart rows, but we only add [nodesInViewport] to
+    // the widget tree.
+    _buildFlameChartElements();
+
+    final nodesInViewport = <FlameChartNode>[];
+    for (var row in rows) {
+      for (var node in row.nodes) {
+        final fitsHorizontally = node.rect.right >= scrollOffsetX &&
+            node.rect.left - scrollOffsetX <= constraints.maxWidth;
+        final fitsVertically = node.rect.bottom >= scrollOffsetY &&
+            node.rect.top - scrollOffsetY <= constraints.maxHeight;
+        if (fitsHorizontally && fitsVertically) {
+          nodesInViewport.add(node);
+        }
+      }
+    }
+    return nodesInViewport;
+  }
+
+  void _buildFlameChartElements() {
+    _resetColorOffsets();
+
+    rows
+      ..clear()
+      ..addAll(List.generate(
+        widget.data.uiEventFlow.depth + widget.data.gpuEventFlow.depth,
+        (i) => FlameChartRow(nodes: [], index: i),
+      ));
+    final int frameStartOffset = widget.data.time.start.inMicroseconds;
+
+    double getTopForRow(int row) {
+      // This accounts for the section spacing between the UI events and the GPU
+      // events.
+      final additionalPadding =
+          row >= gpuSectionStartRow ? sectionSpacing : 0.0;
+      return (row * rowHeightWithPadding + topOffset + additionalPadding)
+          .toDouble();
+    }
+
+    // Pixels per microsecond in order to fit the entire frame in view.
```

Comment:
Note that in Flutter, layouts are computed in dp.

More about this in [devicePixelRatio](https://api.flutter.dev/flutter/dart-ui/Window/devicePixelRatio.html)

Reply:
I don't think this should be an issue as long as `widget.startingContentWidth` (which is passed in with the value `constraints.maxWidth`) is in the same dp as the width/height values we give each flame chart node.

Reply:
Dave I think you had a typo. Flutter layouts are computed in logical pixels not device pixels which is a very good thing.
If they were in device pixels we would have to do a ton of normalizing everywhere to convert back to the appropriate # of device pixels to keep the ui at consistent sizes.

Reply:
> Dave I think you had a typo. Flutter layouts are computed in logical pixels not device pixels which is a very good thing.
> If they were in device pixels we would have to do a ton of normalizing everywhere to convert back to the appropriate # of device pixels to keep the ui at consistent sizes.

The abbreviation for logical pixels is dp (in contrast with px).  The term comes from Android.

https://stackoverflow.com/questions/2025282/what-is-the-difference-between-px-dip-dp-and-sp